File: Golang/intensity-segments.go
package main

import (
	"fmt"
	"sort"
)

// IntensitySegments represents a structure to manage intensity by segments.
// The segments map stores intensity values at specific points.
type IntensitySegments struct {
	segments map[int]int
}

// NewIntensitySegments creates a new instance of IntensitySegments with
// an initial segment value of 0 from -Infinity.
func NewIntensitySegments() *IntensitySegments {
	segments := make(map[int]int)
	segments[-1<<63] = 0 // Represents -Infinity
	return &IntensitySegments{segments: segments}
}

// Add increases the intensity for a given range [from, to) by the specified amount.
func (is *IntensitySegments) Add(from, to, amount int) {
	is.update(from, to, func(currentValue int) int { return currentValue + amount })
}

// Set sets the intensity for a given range [from, to) to the specified amount.
func (is *IntensitySegments) Set(from, to, amount int) {
	is.update(from, to, func(currentValue int) int { return amount })
}

// update modifies the segments between the given range [from, to)
// using the provided update function.
func (is *IntensitySegments) update(from, to int, updateFunc func(int) int) {
	keys := make([]int, 0, len(is.segments)+2)
	for key := range is.segments {
		keys = append(keys, key)
	}
	keys = append(keys, from, to)
	sort.Ints(keys)

	updates := make(map[int]int)
	currentValue := 0
	for _, key := range keys {
		if value, exists := is.segments[key]; exists {
			currentValue = value
		}
		if key >= from && key < to {
			updates[key] = updateFunc(currentValue)
		} else {
			updates[key] = currentValue
		}
	}

	for key, value := range updates {
		is.segments[key] = value
	}
}

// ToString returns a string representation of the current segments.
// The format represents changes in intensity values at specific points.
func (is *IntensitySegments) ToString() string {
	keys := make([]int, 0, len(is.segments))
	for key := range is.segments {
		keys = append(keys, key)
	}
	sort.Ints(keys)

	result := []string{}
	prevValue := 0
	for _, key := range keys {
		if key == -1<<63 {
			continue
		}
		value := is.segments[key]
		if value != prevValue {
			result = append(result, fmt.Sprintf("[%d,%d]", key, value))
			prevValue = value
		}
	}

	return fmt.Sprintf("[%s]", stringJoin(result, ","))
}

// stringJoin joins the elements of a string slice with a specified separator.
func stringJoin(elements []string, sep string) string {
	if len(elements) == 0 {
		return ""
	}
	joined := elements[0]
	for i := 1; i < len(elements); i++ {
		joined += sep + elements[i]
	}
	return joined
}

// Example usage
func TestSegments() {
	segments := NewIntensitySegments()
	fmt.Println(segments.ToString()) // "[]"

	segments.Add(10, 30, 1)
	fmt.Println(segments.ToString()) // "[[10,1],[30,0]]"

	segments.Add(20, 40, 1)
	fmt.Println(segments.ToString()) // "[[10,1],[20,2],[30,1],[40,0]]"

	segments.Add(10, 40, -2)
	fmt.Println(segments.ToString()) // "[[10,-1],[20,0],[30,-1],[40,0]]"

	segmentsN := NewIntensitySegments()
	fmt.Println(segmentsN.ToString()) // "[]"

	segmentsN.Add(10, 30, 1)
	fmt.Println(segmentsN.ToString()) // "[[10,1],[30,0]]"

	segmentsN.Add(20, 40, 1)
	fmt.Println(segmentsN.ToString()) // "[[10,1],[20,2],[30,1],[40,0]]"

	segmentsN.Add(10, 40, -1)
	fmt.Println(segmentsN.ToString()) // "[[20,1],[30,0]]"

	segmentsN.Add(10, 40, -1)
	fmt.Println(segmentsN.ToString()) // "[[10,-1],[20,0],[30,-1],[40,0]]"
}

File: Golang/intensity-segments.js
class IntensitySegments {
    constructor() {
        this.segments = new Map();
        this.segments.set(Number.MIN_SAFE_INTEGER, 0); // Represents -Infinity
    }

    /**
     * Adds an intensity value to the range [from, to).
     * @param {number} from - The start of the range (inclusive).
     * @param {number} to - The end of the range (exclusive).
     * @param {number} amount - The amount to add to the intensity.
     */
    add(from, to, amount) {
        this.update(from, to, (currentValue) => currentValue + amount);
    }

    /**
     * Sets an intensity value for the range [from, to).
     * @param {number} from - The start of the range (inclusive).
     * @param {number} to - The end of the range (exclusive).
     * @param {number} amount - The amount to set as the intensity.
     */
    set(from, to, amount) {
        this.update(from, to, () => amount);
    }

    /**
     * Updates the intensity value for the range [from, to) using the provided update function.
     * @param {number} from - The start of the range (inclusive).
     * @param {number} to - The end of the range (exclusive).
     * @param {function} updateFunc - A function that takes the current value and returns the updated value.
     */
    update(from, to, updateFunc) {
        const keys = Array.from(this.segments.keys());
        keys.push(from, to);
        keys.sort((a, b) => a - b);

        const updates = new Map();
        let currentValue = 0;
        for (let key of keys) {
            if (this.segments.has(key)) {
                currentValue = this.segments.get(key);
            }
            if (key >= from && key < to) {
                updates.set(key, updateFunc(currentValue));
            } else {
                updates.set(key, currentValue);
            }
        }

        for (let [key, value] of updates) {
            this.segments.set(key, value);
        }
    }

    /**
     * Returns a string representation of the current segments.
     * The output format is a list of segment start points and their corresponding intensity values.
     * @returns {string} A string representing the segments.
     */
    toString() {
        const keys = Array.from(this.segments.keys());
        keys.sort((a, b) => a - b);

        let result = [];
        let prevValue = 0;
        for (let key of keys) {
            if (key === Number.MIN_SAFE_INTEGER) continue;
            const value = this.segments.get(key);
            if (value !== prevValue) {
                result.push([key, value]);
                prevValue = value;
            }
        }

        return JSON.stringify(result);
    }
}

// Example usage:
const segments = new IntensitySegments();
console.log(segments.toString()); // "[]"

segments.add(10, 30, 1);
console.log(segments.toString()); // "[[10,1],[30,0]]"

segments.add(20, 40, 1);
console.log(segments.toString()); // "[[10,1],[20,2],[30,1],[40,0]]"

segments.add(10, 40, -2);
console.log(segments.toString()); // "[[10,-1],[20,0],[30,-1],[40,0]]"

const segmentsN = new IntensitySegments();
console.log(segmentsN.toString()); // "[]"

segmentsN.add(10, 30, 1);
console.log(segmentsN.toString()); // "[[10,1],[30,0]]"

segmentsN.add(20, 40, 1);
console.log(segmentsN.toString()); // "[[10,1],[20,2],[30,1],[40,0]]"

segmentsN.add(10, 40, -1);
console.log(segmentsN.toString()); // "[[20,1],[30,0]]"

segmentsN.add(10, 40, -1);
console.log(segmentsN.toString()); // "[[10,-1],[20,0],[30,-1],[40,0]]"
